transformer_engine/pytorch/module/layernorm_linear.py

```diff
@@ -445,6 +454,8 @@ def backward(
                 if not ctx.ub_bulk_dgrad and handle is not None:
                     handle.wait()
                 if not ctx.ub_bulk_wgrad:
+                    if ctx.return_layernorm_output and ctx.return_layernorm_output_gathered:
+                        dgrad = dgrad + grad_outputs[1].view_as(dgrad)
```

Comment:
Shouldn't we do this addition only when we did perform the AG on the forward side? I.e. the case for ColumnParallel execution + SP

Reply:
Yes , we perform AG in forward when `ctx.return_layernorm_output_gathered` is true, so i think this line is correct?

Reply:
Oh i see! thanks for catching this

Reply:
is it sufficient to check for `parallel_mode == "column" and sequence_parallel` in this line?

Reply:
So something like the above suggested changes (untested)

Reply:
Yes that would also work!

Reply:
oh wait, but on line 453, it already checks `if ctx.parallel_mode == "column" and ctx.sequence_parallel:`, so i think we are covered

Reply:
Right and in that case, the addition that should happen in [L552](https://github.com/cuichenx/TransformerEngine/blob/chcui/gathered_layernorm_output/transformer_engine/pytorch/module/layernorm_linear.py/#L552) will be omitted because we are setting `ctx.return_layernorm_output_gathered`  despite not gathering any input.